### ISSUE TITLE
Simple example for working with workspace with multiple binaries

### ISSUE THREAD

**Issue by cameron1024** [nix-community/naersk#272]
Apologies if this is documented somewhere, I had trouble finding it.

I've got a pretty simple repo with 2 binary packages:
```
.
├── Cargo.toml
└── crates
    ├── bar
    │   ├── Cargo.toml
    │   └── src
    │       └── main.rs
    └── foo
        ├── Cargo.toml
        └── src
            └── main.rs
```
and a `flake.nix` that looks like this:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs";
    flake-utils.url = "github:numtide/flake-utils";
    rust-overlay.url = "github:oxalica/rust-overlay";
    naersk.url = "github:nix-community/naersk";
  };

  outputs = { self, nixpkgs, flake-utils, rust-overlay, ... } @ inputs:
    flake-utils.lib.eachDefaultSystem
      (
        system:
        let
          pkgs = import nixpkgs {
            inherit system;
            overlays = [ (import rust-overlay) ];
          };
          naersk = pkgs.callPackage inputs.naersk {};
        in

        {

          packages.foo = naersk.buildPackage {
            name = "foo";
            src = ./.;
          };

          packages.bar = naersk.buildPackage {
            name = "bar";
            src = ./.;
          };

        }

      );
}
```
If I run `nix run .#foo`, I get the following:
```
error: unable to execute '/nix/store/1wgkff4jn798sgrl6fpzx6fdk2z5ydji-foo-unknown/bin/foo-unknown': No such file or directory
```
I've tried adding `targets = ["foo"]` to the call to `buildPackage`, but that doesn't help.

This seems like a really common setup, but I'm struggling to make it work. Are there examples anywhere that show this behaviour?

**Comment by Patryk27** [comment]
So, there are (at least) two options here:

```nix
{ ... }: {
  packages = {
    bar = naersk'.buildPackage {
      src = ./.;
      cargoBuildOptions = opts: opts ++ [ "--package bar" ];
    };

    foo = naersk'.buildPackage {
      src = ./.;
      cargoBuildOptions = opts: opts ++ [ "--package foo" ];
    };
  };
}
```

... or:

```nix
{ ... }: {
  packages =
    let
      pkg = naersk'.buildPackage {
        src = ./.;
      };

    in
    {
      bar = pkgs.runCommand "bar" { } ''
        mkdir -p $out/bin
        ln -s ${pkg}/bin/bar $out/bin
      '';

      foo = pkgs.runCommand "foo" { } ''
        mkdir -p $out/bin
        ln -s ${pkg}/bin/foo $out/bin
      '';
    };
}
```

In the first approach Naresk is called twice, once with `cargo build --package bar` and once with `cargo build --package foo`.

In the second approach Naersk is called just once and later separate expressions are used to "slice" the original output into smaller derivations.

While both approaches should generate the same output, they behave a bit differently during the build-time:

- if you're planning on using all of the generated binaries, and you've got some library-crates in your workspace, then using the second approach might end up in somewhat faster compilation times -- that's because when Cargo is told to compile the entire workspace at once, rustc can re-use artifacts from compiling `bar` in speeding-up compiling `foo` (which it cannot do otherwise, because those artifacts are not kept in the Nix store and so calling `cargo build --package bar` separately from `cargo build --package foo` won't re-use them),

- but, if you're planing on using only some of the generated binaries (or you care about having faster incremental builds), then using the first approach might be faster, because Cargo won't have to compile all of the binaries all of the time.

tl;dr overall I'd probably go with the first approach, since it feels "less hacky" and the potential speedup might not be worth people scratching their head, thinking `what ln -s, huh`

**Comment by cameron1024** [comment]
Hmm, I've tried the first snippet, but it gives me a very similar error, except it's not finding `/nix/store/.../rust-workspace-unknown`. 

As for caching, I usually just use `nix develop` to provide a devshell, then run `cargo build`/`cargo test`/etc as usual. Nix builds are for CI and distribution, where incremental compilation isn't too important.

But I'd definitely prefer the first option, since my team (myself included) is not super familiar with nix, and would like to keep the nix code as simple as possible, and I agree that the `ln -s` stuff is probably confusing.

For context, the full error is:
```
warning: Git tree '/home/cameron/projects/workspace' is dirty
error: unable to execute '/nix/store/2796ra1ir9wzig6rc8mr456jydalindq-rust-workspace-unknown/bin/rust-workspace-unknown': No such file or directory
```

Thanks for the response :grin:

**Comment by Patryk27** [comment]
Hi, sorry for late answer - time flies fast 😅 

Actually, the solution to your problem is pretty simple - using `pname` instead of `name`!

It might feel a bit random, but it stems from a convention used by Nix (well, Nixpkgs) - `name` is something like `bar-1.0.0`, and it's usually generated automatically from `pname` (which stands for "package name") and `version`; and so if you overwrite `name` with something that doesn't follow the `name-version` format, wild things can happen.

Lemme know if it works correctly now 🙂

**Comment by cameron1024** [comment]
Ahh, that's worked perfectly, much cleaner :grin: Thanks for looking at this
